RealmSwift/Tests/RealmTests.swift

```diff
@@ -979,4 +979,413 @@ class RealmTests: TestCase {
+            realm.create(SwiftStringObject.self, value: ["string"])
+            asyncComplete.fulfill()
+        }
+
+        try! realm.write {
+            realm.create(SwiftStringObject.self, value: ["string 2"])
+        }
+
+        waitForExpectations(timeout: 1, handler: nil)
+        XCTAssertEqual(1, realm.objects(SwiftStringObject.self).count)
+        XCTAssertEqual("string 2", realm.objects(SwiftStringObject.self).first?.stringCol)
+    }
+
+    func testAsyncTransactionWriteWithSync() {
+        let realm = try! Realm()
+        let asyncComplete = expectation(description: "async transaction complete")
+
+        XCTAssertEqual(0, realm.objects(SwiftStringObject.self).count)
+
+        try! realm.write {
+            realm.create(SwiftStringObject.self, value: ["string"])
+        }
+
+        realm.beginWrite()
+        realm.create(SwiftStringObject.self, value: ["string 2"])
+        realm.commitAsyncWrite {
+            asyncComplete.fulfill()
+        }
+
+        waitForExpectations(timeout: 1, handler: nil)
+        XCTAssertEqual(2, realm.objects(SwiftStringObject.self).count)
+    }
+
+    func testAsyncTransactionMixedWithSync() {
+        let realm = try! Realm()
+        let asyncComplete = expectation(description: "async transaction complete")
+
+        XCTAssertEqual(0, realm.objects(SwiftStringObject.self).count)
+
+        try! realm.writeAsync { _ in
+            realm.create(SwiftStringObject.self, value: ["string"])
+        }
+
+        try! realm.writeAsync { _ in
+            realm.create(SwiftStringObject.self, value: ["string 2"])
+        } _: {
+            asyncComplete.fulfill()
+        }
+
+        realm.beginWrite()
+        realm.create(SwiftStringObject.self, value: ["string 3"])
+        try! realm.commitWrite()
+
+        waitForExpectations(timeout: 1, handler: nil)
+        XCTAssertEqual(3, realm.objects(SwiftStringObject.self).count)
+    }
+
+    func testAsyncTransactionMixedWithCancelledSync() {
+        let realm = try! Realm()
+        let asyncComplete = expectation(description: "async transaction complete")
+
+        XCTAssertEqual(0, realm.objects(SwiftStringObject.self).count)
+
+        try! realm.writeAsync { _ in
+            realm.create(SwiftStringObject.self, value: ["string"])
+        }
+
+        try! realm.writeAsync { _ in
+            realm.create(SwiftStringObject.self, value: ["string 2"])
+        } _: {
+            asyncComplete.fulfill()
+        }
+
+        realm.beginWrite()
+        realm.create(SwiftStringObject.self, value: ["string 3"])
+        realm.cancelWrite()
+
+        waitForExpectations(timeout: 1, handler: nil)
+        XCTAssertEqual(2, realm.objects(SwiftStringObject.self).count)
+    }
+
+    func testAsyncTransactionChangeNotification() {
+        let realm = try! Realm()
+        let asyncComplete = expectation(description: "async transaction complete")
+        asyncComplete.expectedFulfillmentCount = 4
```

Comment:
Why is the .update case expected to be called 4 times when two writes are made?

Reply:
2 for write, 2 for update notification. I'll add another expectation for update to make it clear